GameState serialization call

datanoh opened the pull request:
Added two functions to communication.js to get the serialized game state. I'm not that clear on what kind of syntax is required on the JS side, do either of these look correct to you?

New functions are getGameStateOld and getGameStateNew. Presumably we don't need both :)

datanoh commented:
@andrewd18 - What's the long-term on this? It occurs to me that, as it's currently built, it's a very easy way for either player to see a ton of information they shouldn't have available to them

JeBuSBrian commented:
> @andrewd18 - What's the long-term on this? It occurs to me that, as it's currently built, it's a very easy way for either player to see a ton of information they shouldn't have available to them

I would suggest not allowing local serialization of game state.

Having said that, I know from my own stalled attempts to create an online version of the games that it's easier said than done. State needs to be transmitted between client and server.

So, at the end of the day, my practical suggestion would be to do your best to serialize only what is knowable by that player. That should be fairly doable, as what is knowable is already what the player can actually see on the screen.

Everything else can be serialized server-side, if necessary. If at all possible, nothing should be serialized client-side.

andrewd18 commented:
`GameStateNew` is the preferred approach using the modern, standard [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). All of the error handling is done right inside that function, and the function is async so it doesn't block the rest of the client if the response is slow.

> I would suggest not allowing client-side serialization of game state.

As you mention, some state must be transmitted or else the client shows an empty page.

> So, at the end of the day, my practical suggestion would be to do your best to serialize only what is knowable by that player. That should be fairly doable, as what is knowable is already what the player can actually see on the screen.

I believe this is the long-term plan. At the moment the risk is low because we don't have players, much less players who would open developer tools to get hidden information about the opponent. Also, there are already plenty of other gaping security holes in GEMP that a crafty player could use, including full deck lists for all decks, editing people's decks before a game started, sending false game events to discard opponents' hands or make them concede, etc. A security pass is going to be necessary at some point.

For some context, the reasons I asked for this feature were:

1. The current design has the client maintaining its own, separate state of the current game in browser memory. This is a risk for desyncs from the server.
2. When you close and reload the page, the server sends every single card play event in a specific order to the client. This is a risk for desyncs from the server, and it would be much faster if I could just parse a single, base gamestate object.
3. React.js will be used to build the client UI going forward. It works on a top-down, tree-based, data driven approach that rapidly parses content from a data provider (a single server or a GraphQL server in big projects) to break that data up and hand it to the appropriate UI elements. This is much easier to maintain than dozens of individual calls to update UI elements individually, but does require all the data in one data set.  If we need to split this up into smaller UI calls for security concerns, then so be it, but I think we can tackle that down the road.

andrewd18 commented:
@datanoh Yes, I eventually want gamestate to only reveal what the player should know. This means the gamestate call should compare against a player session token created at auth login which the UI passes to the server along with its requests.

datanoh commented:
@andrewd18 - Can you check this out?

I'm passing tests on the server side for both the full GameState and the player-specific GameStateView objects. The communication with the client is going through GameRequestHandler.java:

https://github.com/trekcc-ops/gemp-stccg/blob/8e551cf8f38e5e6677e3839071e400c873adf7a6/gemp-module/gemp-stccg-server/src/main/java/com/gempukku/stccg/async/handler/GameRequestHandler.java#L62-L69

The getGameState method seems to be working fine, getGameStateForPlayer not so much. But it's possible I'm missing something on the client side that's needed to handle it. (I added lines 66-69 in hopes that I could brute force the player-specific game state without being logged in as a player, but no luck.)

Related news - I'm not able to debug the server-side code anymore. I think the recent Docker changes broke something in my setup. I'll open an issue for it.

datanoh commented:
@andrewd18 - After the fix you provided for debugging, I was able to get this working. I've got a bit of cleanup to do with it. Will reach out when it's in a better state for review.

datanoh commented:
@andrewd18 - Take a look when you have some time. The client now has access to four different game state views (replace XX in the links below with the game id number):
- /gemp-stccg-server/game/XX/gameState/player <- This is the "player-specific" game state that will be sent to players. It is a limited view, and doesn't include information that we don't think a player should have access to (like what order the cards in their draw deck are in, cards in opponent's hand, etc.)
- /gemp-stccg-server/game/XX/gameState/admin <- This is the full game state. It will only be available to admin users.
- /gemp-stccg-server/game/XX/gameState/player1 (and player2) <- This is a copy of both of the player-specific game states, only available to admin users. Tempted to get rid of these, but wanted to run them by you first.

I updated the documentation in gamestate-example.hjson to reflect the new information in these files.

I also added Issue #148. This includes some documentation about the current GameEvent usage. I've added a new Game Event - "GAME_STATE_CHECK". It's an empty event, and only exists to tell the client to request the JSON game state. It is currently sent when cards are stopped, or when a card is placed on a mission.

andrewd18 commented:
Confirming I can get the player specific JSON. Need to set up an admin user and make a deck, etc. to test the other states.
